### ISSUE TITLE
Explicit count for each kind of import

### ISSUE THREAD

**Issue by binji** [WebAssembly/design#882]
@sbc100 were discussing object linking today (see https://github.com/WebAssembly/tool-conventions/blob/master/Linking.md).

For a fast linker, you'll want to parse as little as possible, and rely on the relocations to fix indexing. But it seems as though parsing the import section is necessary, as you can't know without parsing how many of each kind of import there are, and you'll need to know this number to fix up references.

One simple fix would be to include the count of each import kind in the section header. OTOH, the import section is likely to be small, so maybe this is not worth doing. Thoughts?

**Comment by jfbastien** [comment]
I'm not sure I like the extra validation we'd need to do. It's tiny, but has correctness implications on the spec and implementation, whereas dynamic linking doesn't have these downsides and isn't bottle-necked compared to the rest of the compilation.

**Comment by lukewagner** [comment]
I don't quite follow: are you saying there is a problem, even given [`Module.imports`](https://github.com/WebAssembly/design/blob/master/JS.md#webassemblymoduleimports), which includes the `kind` of the import?

**Comment by dschuff** [comment]
Well this would be more for a (static) linker, an external tool (rather than JS code) which would consume N wasm files and produce a new one.

**Comment by rossberg-chromium** [comment]
Wouldn't a linker need the types of the imports? I don't see how just having the counts helps.

**Comment by lukewagner** [comment]
Oh, *static* linker.  Wouldn't such a linker also though need to decode the entire imports/exports section (to get all the names to link)?

**Comment by AndrewScheidecker** [comment]
This would be somewhat mitigated by my proposal in #783 to merge the module-scoped index spaces. It would make it possible to resolve an index to an import or definition knowing only the number of declarations in each section, where you currently need to parse the imports section into a list of imports of each kind. You would still need to deserialize the import section to get type information, but that seems unavoidable.

**Comment by dschuff** [comment]
It's true that you might need to parse the import section for the symbol-resolution phase (at least for object files in the link). For archives though, you might not (assuming you used an archive symbol table).
For the index fixups in the text section, you don't need the types, because the function index space is unified. For e.g. creating a new function signature section you'd probably also generate a new one from the internal symbol table. At least for the simple cases. Not sure about some of the more complex things though (e.g. comdats). Also come to think of it I'm not sure an archive symbol table is sufficiently expressive to completely avoid parsing the imports during symbol resolution.

**Comment by binji** [comment]
Sorry, what I meant was more of a object merge operation, rather than a link. You are right that you'll need the import section to actually link. I brought this up because I've noticed dealing with the separate index spaces in the import section was a bit of annoyance in wabt too. But like @AndrewScheidecker says, if we were to simplify this, it might be better to just unify the index spaces.

**Comment by rossberg-chromium** [comment]
@binji, ...or conversely, split the import section like the rest. I also found the mixture tedious on a few occasions, in both the spec interpreter and the formalisation. For example, there is no easy way to tell whether a given index is an import or not (which merging the namespaces would not really fix).

But overall, I'm not sure it's worth changing. I actually played with a similar change in the spec interpreter's AST a while ago, but it just led to complications elsewhere.
